lib/BuildSystem/LaneBasedExecutionQueue.cpp

```diff
@@ -126,14 +152,42 @@ class LaneBasedExecutionQueue : public BuildExecutionQueue {
     for (unsigned i = 0; i != numLanes; ++i) {
       lanes[i]->join();
     }
+
+    if (killAfterTimeoutThread) {
+      stopKillingCondition.notify_all();
+      killAfterTimeoutThread->join();
+    }
   }
 
   virtual void addJob(QueueJob job) override {
+    if (cancelled && !job.isSentinel()) {
+      // FIXME: We should eventually raise an error here as new work should not be enqueued after cancellation
+      return;
+    }
+
     std::lock_guard<std::mutex> guard(readyJobsMutex);
     readyJobs.push_back(job);
     readyJobsCondition.notify_one();
   }
 
+  virtual void cancelAllJobs() override {
+    // We have already been cancelled
+    if (killAfterTimeoutThread != nullptr) {
+      return;
+    }
+
+    cancelled = true;
```

Comment:
This method isn't thread safe, I think what it can do instead is compare-and-swap on cancelled, and then if it won the CAS *then* it can adjust `killAfterTimeoutThread`.